### DISCUSSION TITLE
opentelemetry-instrumentation-faraday TracerMiddleware position in RackBuilder

### DISCUSSION THREAD

**Discussion by ib-drct-dev** [open-telemetry/opentelemetry-ruby#955] [General]
I use faraday opentelemetry instrumentation with a logger middleware. 

```
conn = Faraday.new(conn_options) do |builder|
  builder.request  :url_encoded
  builder.response :logger
  builder.adapter :net_http
end
```

Logs should be tagged with trace_id and span_id.

But there is a problem: OpenTelemetry::Instrumentation::Faraday::Patches::RackBuilder inserts TracerMiddleware before an adapter. Thus a logger is left out of Faraday's request scope and tagged with a wrong span (it uses parent's scope).

https://github.com/open-telemetry/opentelemetry-ruby/blob/main/instrumentation/faraday/lib/opentelemetry/instrumentation/faraday/patches/rack_builder.rb#L14

Would you accept as a solution to insert TracerMiddleware to the top of a middleware stack?

```
module OpenTelemetry
  module Instrumentation
    module Faraday
      module Patches
        module RackBuilder
          def adapter(*args)
            insert(0,
              ::Faraday::Middleware.lookup_middleware(:open_telemetry)) unless @handlers.any? do |handler|
              handler.klass == Faraday::Middlewares::TracerMiddleware
            end

            super
          end
        end
      end
    end
  end
end
```

**Comment by ahayworth** [comment]
Hi @ib-drct-dev - this was actually just brought up a few weeks ago (albeit for a different reason). At the time, @fbogsany mentioned that if you insert the tracing middleware yourself manually, then our patches respect the position you chose - see [this comment](https://github.com/open-telemetry/opentelemetry-ruby/pull/935#issuecomment-916980928) for more details.

Would that solution work for you?

**Reply by ib-drct-dev** [reply]
Yeah, It will. But it would be a bit cumbersome for my case because then I have to config 20-30 connections in my project.
Actually, I've already monkey patched faraday instrumentation for my case.

But, maybe it's also an option to update insert logic the next way: check if there is a logger middleware and if so then insert tracing middleware before it?

**Reply by fbogsany** [reply]
> But, maybe it's also an option to update insert logic the next way: check if there is a logger middleware and if so then insert tracing middleware before it?

I think trying to auto-detect every scenario people might want and then inserting the middleware in exactly the right position is going to be a maintenance nightmare. I think the right approach is to add a config option to indicate which middleware (if any) the tracing middleware should be inserted before. In some sense, we're working around the awkwardness of Faraday configuration for these sorts of cases, where users have many connections to configure.

**Reply by ib-drct-dev** [reply]
> I think trying to auto-detect every scenario people might want and then inserting the middleware in exactly the right position is going to be a maintenance nightmare. 

Yeah, I see. Alright, thank you for answers!
